### PR TITLE
ExUnit: Rename variables and refactor config and loop 

### PR DESCRIPTION
**Pull request by eksperimental**
This is a preliminary PR in order to implement the --max-failures feature in ExUnit

It only renames arguments, variables, and reorganizes code.
As requested in https://github.com/elixir-lang/elixir/pull/8247#issuecomment-426530597

### CODE REVIEW

File: lib/ex_unit/lib/ex_unit/runner.ex [hunk 1]
```diff
@@ -2,14 +2,18 @@ defmodule ExUnit.Runner do
   @moduledoc false
 
   alias ExUnit.EventManager, as: EM
+  alias ExUnit.RunnerStats
```
Comment (eksperimental) [inline lib/ex_unit/lib/ex_unit/runner.ex]: This alias is used in several instances in the --max-failures implementation

File: lib/ex_unit/lib/ex_unit/runner.ex [hunk 2]
```diff
@@ -229,13 +231,13 @@ defmodule ExUnit.Runner do
...
 
-  defp run_test(config, %{tags: tags} = test, context) do
+  defp run_test(%{tags: tags} = test, context, config) do
     EM.test_started(config.manager, test)
 
     test =
       if is_nil(test.state) do
-        capture_log? = Map.get(tags, :capture_log, config.capture_log)
-        run_test_with_capture_log(capture_log?, config, test, Map.merge(tags, context))
+        capture_log = Map.get(tags, :capture_log, config.capture_log)
```
Comment (eksperimental) [inline lib/ex_unit/lib/ex_unit/runner.ex]: I have removed the `?` from this variables, as it is not only boolean values that will hold

File: lib/ex_unit/lib/ex_unit/runner.ex [hunk 3]
```diff
@@ -335,6 +344,10 @@ defmodule ExUnit.Runner do
 
   ## Helpers
 
+  defp generate_test_seed(%ExUnit.Test{module: module, name: name}, %{seed: seed}) do
+    :rand.seed(@rand_algorithm, {:erlang.phash2(module), :erlang.phash2(name), seed})
+  end
+
```
Comment (eksperimental) [inline lib/ex_unit/lib/ex_unit/runner.ex]: this was moved here under the Helper section

File: lib/ex_unit/lib/ex_unit/runner.ex [hunk 4]
```diff
@@ -187,7 +190,6 @@ defmodule ExUnit.Runner do
           {test_module, [], []}
       end
 
-    timeout = get_timeout(%{}, config)
```
Comment (eksperimental) [inline lib/ex_unit/lib/ex_unit/runner.ex]: moved to the top of the function, where we define variables

File: lib/ex_unit/lib/ex_unit/runner.ex [hunk 5]
```diff
@@ -2,14 +2,18 @@ defmodule ExUnit.Runner do
...
+  @type manager :: ExUnit.EventManager.manager()
+
   def run(opts, load_us) do
     {:ok, manager} = EM.start_link()
-    {:ok, stats} = EM.add_handler(manager, ExUnit.RunnerStats, opts)
+    {:ok, stat_sup} = EM.add_handler(manager, RunnerStats, opts)
     {opts, config} = configure(manager, opts)
-
+    config = Map.put(config, :stat_sup, stat_sup)
+    config = Map.put(config, :top_pid, self())
```
Comment (eksperimental) [inline lib/ex_unit/lib/ex_unit/runner.ex]: this is the only addition to adds "a feature", we store the top_pid and stat_sup pids, and we avoid to add an extra argument to functions. Maybe there's a better name for top_pid?

File: lib/ex_unit/lib/ex_unit/runner.ex [hunk 6]
```diff
@@ -264,21 +267,27 @@ defmodule ExUnit.Runner do
...
 
-  defp generate_test_seed(%{seed: seed}, %ExUnit.Test{module: module, name: name}) do
-    :rand.seed(@rand_algorithm, {:erlang.phash2(module), :erlang.phash2(name), seed})
+    exec_on_exit(test, test_pid, timeout)
   end
 
-  defp receive_test_reply(test, test_pid, test_ref, timeout) do
+  defp receive_test_reply(
+         test,
+         %{pid: test_pid, ref: test_ref, timeout: timeout} = options
```
Comment (josevalim) [inline lib/ex_unit/lib/ex_unit/runner.ex]: Please move the pattern to inside the function.

File: lib/ex_unit/lib/ex_unit/event_manager.ex [hunk 7]
```diff
@@ -2,6 +2,8 @@ defmodule ExUnit.EventManager do
   @moduledoc false
   @timeout 30000
 
+  @type manager :: {supervisor_manager :: pid, event_manager :: pid}
```
Comment (josevalim) [inline lib/ex_unit/lib/ex_unit/event_manager.ex]: Please make it `@typep`.

File: lib/ex_unit/lib/ex_unit/runner.ex (hunk 5 above)
Comment (josevalim) [reply lib/ex_unit/lib/ex_unit/runner.ex]: So lets make sure that we add those when we create the map and not afterwards. Otherwise we get the impression the map is dynamic but it keys are all static (i.e. known upfront). Let's also call those `stats_pid` and `runner_pid`. After all, we don't know that the pid for stats is a sup, that would be leaking implementation detail.

**Comment by josevalim** [comment]
Tks @eksperimental. What is the reason to change the order of config?

**Comment by josevalim** [comment]
Before this PR, the only function that received `config` as last argument was `get_timeout`. When we change the argument order, we are generating a bunch of churn for not clear benefit. Unless we have a strong reason for doing so, I would recommend to revert the change. There is nothing that says the `state` like argument has to be the last one and it is not a convention we particularly follow for internal functions, especially when a behaviour is *not* involved.

File: lib/ex_unit/lib/ex_unit/runner_stats.ex [hunk 8]
```diff
@@ -4,8 +4,8 @@ defmodule ExUnit.RunnerStats do
   use GenServer
   alias ExUnit.{FailuresManifest, Test, TestModule}
 
-  def stats(pid) do
-    GenServer.call(pid, :stats, :infinity)
+  def stats(sup) do
```
Comment (fertapric) [inline lib/ex_unit/lib/ex_unit/runner_stats.ex]: I would use `pid` here

**Comment by eksperimental** [comment]
> Before this PR, the only function that received `config` as last argument was `get_timeout`. So if the goal is consistency, I would change `get_timeout`.
> 
> When we change the argument order, we are generating a bunch of churn for not clear benefit. Unless we have a strong reason for doing so, I would recommend to revert the change. There is nothing that says the "state-like" argument has to be the last one and it is not a convention we particularly follow for internal functions, especially when a behaviour is _not_ involved.

I have never considered `config` a state-like argument. It has always looked to me as an `options` argument, and that's the reason why it was moved to the end of the arg list.
Looking into it, I realized that the only state value was `modules`, and the pids and the manager, but those while not provided by the user, they never change so they could be considered part of config.
I have refactored the code. Now it does more than just renaming variables, but I think it defines the line in what's state and what's a configuration option

File: lib/ex_unit/lib/ex_unit/runner_stats.ex (hunk 8 above)
Comment (eksperimental) [reply lib/ex_unit/lib/ex_unit/runner_stats.ex]: fixed

File: lib/ex_unit/lib/ex_unit/event_manager.ex (hunk 7 above)
Comment (eksperimental) [reply lib/ex_unit/lib/ex_unit/event_manager.ex]: done

File: lib/ex_unit/lib/ex_unit/runner.ex (hunk 6 above)
Comment (eksperimental) [reply lib/ex_unit/lib/ex_unit/runner.ex]: done

File: lib/ex_unit/lib/ex_unit/runner.ex (hunk 5 above)
Comment (eksperimental) [reply lib/ex_unit/lib/ex_unit/runner.ex]: done

**Comment by eksperimental** [comment]
please do a second review if you like @fertapric, @josevalim; as I have updated the PR

File: lib/ex_unit/lib/ex_unit/runner.ex [hunk 9]
```diff
@@ -5,45 +5,46 @@ defmodule ExUnit.Runner do
 
   @rand_algorithm :exs1024
 
-  def run(opts, load_us) do
-    {:ok, manager} = EM.start_link()
-    {:ok, stats} = EM.add_handler(manager, ExUnit.RunnerStats, opts)
-    {opts, config} = configure(manager, opts)
+  @type manager :: ExUnit.EventManager.manager()
+
+  def run(load_time, opts) when (is_integer(load_time) or is_nil(load_time)) and is_list(opts) do
```
Comment (eksperimental) [inline lib/ex_unit/lib/ex_unit/runner.ex]: Not sure if its a wise choice to change the argument order here for the sake of consistency.
Could it affect third party libraries? 
WDYT @josevalim?

Comment (josevalim) [reply lib/ex_unit/lib/ex_unit/runner.ex]: Let's please just avoid those changes generally. The benefits are very little over the history change and potential breakage.

File: lib/ex_unit/lib/ex_unit/runner.ex [hunk 10]
```diff
@@ -5,45 +5,46 @@ defmodule ExUnit.Runner do
...
+    %{
       capture_log: opts[:capture_log],
       exclude: opts[:exclude],
       include: opts[:include],
       manager: manager,
       max_cases: opts[:max_cases],
       only_test_ids: opts[:only_test_ids],
+      runner_pid: runner_pid,
       seed: opts[:seed],
-      modules: :async,
```
Comment (eksperimental) [inline lib/ex_unit/lib/ex_unit/runner.ex]: so now, `modules` is passed around by the loop function, as there is no need to keep it in config.

File: lib/ex_unit/lib/ex_unit/runner.ex [hunk 11]
```diff
@@ -5,45 +5,46 @@ defmodule ExUnit.Runner do
 
   @rand_algorithm :exs1024
 
-  def run(opts, load_us) do
-    {:ok, manager} = EM.start_link()
-    {:ok, stats} = EM.add_handler(manager, ExUnit.RunnerStats, opts)
-    {opts, config} = configure(manager, opts)
+  @type manager :: ExUnit.EventManager.manager()
```
Comment (josevalim) [inline lib/ex_unit/lib/ex_unit/runner.ex]: We can't do this because the type is privat. Are we even using it?

Comment (eksperimental) [reply lib/ex_unit/lib/ex_unit/runner.ex]: I have the feeling I do in the main PR.
I will remove it for now and add it back if needed

File: lib/ex_unit/lib/ex_unit/runner.ex [hunk 12]
```diff
@@ -5,45 +5,46 @@ defmodule ExUnit.Runner do
...
+
+  def run(load_time, opts) when (is_integer(load_time) or is_nil(load_time)) and is_list(opts) do
+    opts = normalize_opts(opts)
 
+    {:ok, manager} = EM.start_link()
+    {:ok, stats_pid} = EM.add_handler(manager, ExUnit.RunnerStats, opts)
+    config = configure(opts, manager, self(), stats_pid)
     :erlang.system_flag(:backtrace_depth, Keyword.fetch!(opts, :stacktrace_depth))
 
-    {run_us, _} =
```
Comment (josevalim) [inline lib/ex_unit/lib/ex_unit/runner.ex]: I think we are losing important information here, which is that the runtime is in `us`.

File: lib/ex_unit/lib/ex_unit/runner.ex [hunk 13]
```diff
@@ -54,32 +55,34 @@ defmodule ExUnit.Runner do
     |> Keyword.put(:include, include)
   end
 
-  defp loop(%{modules: :async} = config, taken) do
+  defp loop(modules \\ :async, taken \\ 0, config)
```
Comment (josevalim) [inline lib/ex_unit/lib/ex_unit/runner.ex]: I personally find an antipattern to use default arguments for private functions. Can we please pass all arguments the time we call it? Thanks.

**Comment by josevalim** [comment]
@eksperimental I have added some comments. Can we please keep this PR focused? In your latest commits there are new changes compared to the earlier commit, so we have to review the whole PR again, and probably another pass later on once you do your changes again. It is generally preferred for us to keep PRs smalls and focused.

File: lib/ex_unit/lib/ex_unit/runner.ex (hunk 11 above)
Comment (josevalim) [reply lib/ex_unit/lib/ex_unit/runner.ex]: It shouldn't be necessary in the main PR either because we don't use spec for private functions or private functionality and everything here is private.

File: lib/ex_unit/lib/ex_unit/runner.ex (hunk 12 above)
Comment (fertapric) [reply lib/ex_unit/lib/ex_unit/runner.ex]: Agree with @josevalim. I would use `usec` though.

File: lib/ex_unit/lib/ex_unit/runner.ex [hunk 14]
```diff
@@ -90,52 +93,50 @@ defmodule ExUnit.Runner do
...
 
-  defp run_module(config, pid, module) do
+  defp run_module(module, config) do
     test_module = module.__ex_unit__()
     EM.module_started(config.manager, test_module)
 
     # Prepare tests, selecting which ones should
     # run and which ones were skipped.
-    tests = prepare_tests(config, test_module.tests)
+    tests = prepare_tests(test_module.tests, config)
```
Comment (josevalim) [inline lib/ex_unit/lib/ex_unit/runner.ex]: Please revert everywhere we change the order of arguments. It is an unnecessary change that complicates history with no benefit. We will keep `config` as first argument. Thanks.

File: lib/ex_unit/lib/ex_unit/runner.ex (hunk 12 above)
Comment (josevalim) [reply lib/ex_unit/lib/ex_unit/runner.ex]: So I would go with `runtime_us`.

File: lib/ex_unit/lib/ex_unit/runner.ex [hunk 15]
```diff
@@ -264,21 +266,25 @@ defmodule ExUnit.Runner do
...
+
+    test = receive_test_reply(test, reply_opts)
 
     exec_on_exit(test, test_pid, timeout)
   end
 
-  defp generate_test_seed(%{seed: seed}, %ExUnit.Test{module: module, name: name}) do
-    :rand.seed(@rand_algorithm, {:erlang.phash2(module), :erlang.phash2(name), seed})
-  end
+  defp receive_test_reply(test, opts) do
```
Comment (josevalim) [inline lib/ex_unit/lib/ex_unit/runner.ex]: It should not be called `opts`, it should be called `config`, as it is a structured map.

File: lib/ex_unit/lib/ex_unit/runner.ex [hunk 16]
```diff
@@ -90,52 +93,50 @@ defmodule ExUnit.Runner do
...
       {_pid, :module_finished, _test_case} ->
-        loop(config, taken - 1)
+        loop(modules, taken - 1, config)
     end
   end
 
-  defp spawn_modules(config, modules, taken) do
-    pid = self()
-
+  defp spawn_modules(modules, modules_left, taken, config) do
```
Comment (josevalim) [inline lib/ex_unit/lib/ex_unit/runner.ex]: `remaining_modules` or `modules_remaining` may be a better name then `modules_left`.

File: lib/ex_unit/lib/ex_unit/runner.ex (hunk 15 above)
Comment (eksperimental) [reply lib/ex_unit/lib/ex_unit/runner.ex]: good to know, I didn't know that was the difference

File: lib/ex_unit/lib/ex_unit/runner.ex (hunk 12 above)
Comment (eksperimental) [reply lib/ex_unit/lib/ex_unit/runner.ex]: The reason I changed it its because `run_us` and `load_us` since it has a meaning in English, it was confusing, I have never seen that term before and neither it is used in Elixir core, only one time here.
Took me a while to figure it out.

If I had read `run_usec` I would have figured it out instantly

File: lib/ex_unit/lib/ex_unit/runner.ex [hunk 17]
```diff
@@ -264,21 +262,24 @@ defmodule ExUnit.Runner do
...
         exit(:shutdown)
       end)
 
-    timeout = get_timeout(test.tags, config)
-    test = receive_test_reply(test, test_pid, test_ref, timeout)
+    reply_config = %{
+      pid: test_pid,
+      ref: test_ref,
+      timeout: timeout
+    }
```
Comment (josevalim) [inline lib/ex_unit/lib/ex_unit/runner.ex]: Can we please keep the previous call? It is fine for us to call a function with four arguments. I don't think it makes much sense to build a map just to unpack it right after.

**Comment by josevalim** [comment]
Thanks @eksperimental, one last comment and we should be good to go.

**Comment by whatyouhide** [comment]
Nice job @eksperimental, thanks! 💟

**Comment by whatyouhide** [comment]
Nice job @eksperimental, thanks! 💟

**Comment by eksperimental** [comment]
thank you @whatyouhide,

BTW, there is an error in my implementation when passing the modules which most of them are being lost.
I'm working on it, so please don't merge any more PRs as the tests are passing but very few tests are being run.
sorry for the inconvenience
